File: hdf-converters/converters/hdf-to-oscal-sar/go/tierounding_test.go
package hdftooscalsar

import (
	"os"
	"path/filepath"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// Impact 0.25 is exactly representable and ties at one decimal, where fmt's
// %.1f rounds to even ("0.2") and JavaScript's toFixed rounds away from zero
// ("0.3"). The TypeScript peer asserts the same string on the same fixture.
func TestRiskImpactTextRendersTiesLikeToFixed(t *testing.T) {
	in, err := os.ReadFile(filepath.Join("..", "fixtures", "input", "tie-rounding.json"))
	require.NoError(t, err)
	out, err := ConvertHDFToOSCALSAR(in, "0.1.0")
	require.NoError(t, err)

	assert.Contains(t, string(out), "Impact: 0.3 (", "0.25 at one decimal must round away from zero, as toFixed does")
	assert.NotContains(t, string(out), "Impact: 0.2 (", "round-half-to-even leaked through")
}

File: hdf-converters/converters/hdf-to-oscal-sar/typescript/tie-rounding.test.ts
import { describe, it, expect } from 'vitest';
import * as testhdf from '@mitre/hdf-schema/testhdf';
import { convertHdfToOscalSar } from './converter.js';

// The Go peer asserts this same string on the same shape. Impact 0.25 ties at
// one decimal, where toFixed rounds away from zero and Go's fmt rounds to even.
// Built with the testhdf builder rather than a committed fixture: the document
// asserts nothing about real-world data shape, only about how one number renders.
function tieDocument(): string {
  return JSON.stringify(
    testhdf.doc(
      testhdf.baseline(
        'tie',
        testhdf.req('AC-1', {
          impact: 0.25,
          tags: { nist: ['AC-1'] },
          desc: 'impact 0.25 ties at one decimal',
          status: 'failed',
        }),
      ),
    ),
  );
}

describe('hdf-to-oscal-sar renders exact ties the way Go now does', () => {
  it('rounds impact 0.25 to 0.3 at one decimal', async () => {
    const out = await convertHdfToOscalSar(tieDocument());
    expect(out).toContain('Impact: 0.3 (');
    expect(out).not.toContain('Impact: 0.2 (');
  });
});
